File: Data Structures & Algorithms/implement-queue-using-stacks/submission-0.go
type MyQueue struct {
	list []int
}

func Constructor() MyQueue {
	return MyQueue{[]int{}}
}

func (this *MyQueue) Push(x int) {
	this.list = append(this.list, x)
}

func (this *MyQueue) Pop() int {
	if this.Empty() {
		return -1
	} 
	peek := this.Peek()
	this.list = this.list[1:]
	return peek
}

func (this *MyQueue) Peek() int {
	if this.Empty() {
		return -1
	}
	return this.list[0]
}

func (this *MyQueue) Empty() bool {
	return len(this.list) == 0
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * param2 := obj.Pop();
 * param3 := obj.Peek();
 * param4 := obj.Empty();
 */

File: Data Structures & Algorithms/design-browser-history/submission-0.js
class BrowserHistory {
    /**
     * @constructor
     * @param {string} homepage
     */
    constructor(homepage) {
        this.stack = [homepage];
        this.curr = 0;
    }

    /**
     * @param {string} url
     * @return {void}
     */
    visit(url) {
        this.stack = this.stack.slice(0, this.curr + 1);
        this.stack.push(url);
        this.curr = this.stack.length - 1;
    }

    /**
     * @param {number} steps
     * @return {string}
     */
    back(steps) {
        this.curr = Math.max(0, this.curr - steps);
        return this.stack[this.curr];
    }

    /**
     * @param {number} steps
     * @return {string}
     */
    forward(steps) {
        this.curr = Math.min(this.stack.length - 1, this.curr + steps);
        return this.stack[this.curr];
    }
}

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * var obj = new BrowserHistory(homepage)
 * obj.visit(url)
 * var param_2 = obj.back(steps)
 * var param_3 = obj.forward(steps)
 */
